theories/lra.elpi

```diff
@@ -0,0 +1,498 @@
+
+% [field-mode] is true if we are on a realFieldType,
+% otherwise we are on a realDomainType.
+pred field-mode.
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+% Parse goal (and hypotheses) to extract a realFieldType or realDomainType
+% from (in)equalities it contains
+
+% ring or field structure from a term of type bool
+pred rfstr.bool i:term, o:rf.
+rfstr.bool {{ _ ==> lp:Type }} F :- rfstr.bool Type F.
+rfstr.bool {{ lp:Type ==> _ }} F :- rfstr.bool Type F.
+rfstr.bool {{ ~~ lp:Type }} F :- rfstr.bool Type F.
+rfstr.bool {{ _ && lp:Type }} F :- rfstr.bool Type F.
+rfstr.bool {{ lp:Type && _ }} F :- rfstr.bool Type F.
+rfstr.bool {{ _ || lp:Type }} F :- rfstr.bool Type F.
+rfstr.bool {{ lp:Type || _ }} F :- rfstr.bool Type F.
+rfstr.bool {{ @Order.le _ lp:Ty _ _ }} F :- field-mode,
+  rf->porder F Ty, F = field _.
+rfstr.bool {{ @Order.lt _ lp:Ty _ _ }} F :- field-mode,
+  rf->porder F Ty, F = field _.
+rfstr.bool {{ @Order.le _ lp:Ty _ _ }} R :- not field-mode,
+  rf->porder R Ty, R = ring _.
+rfstr.bool {{ @Order.lt _ lp:Ty _ _ }} R :- not field-mode,
+  rf->porder R Ty, R = ring _.
+
+% ring or field structure from a term of type Prop
+pred rfstr.prop i:term, o:rf.
+rfstr.prop {{ _ -> lp:Type }} F :- rfstr.prop Type F.
+rfstr.prop {{ lp:Type -> _ }} F :- rfstr.prop Type F.
+rfstr.prop {{ iff _ lp:Type }} F :- rfstr.prop Type F.
+rfstr.prop {{ iff lp:Type _ }} F :- rfstr.prop Type F.
+rfstr.prop {{ ~ lp:Type }} F :- rfstr.prop Type F.
+rfstr.prop {{ _ /\ lp:Type }} F :- rfstr.prop Type F.
+rfstr.prop {{ lp:Type /\ _ }} F :- rfstr.prop Type F.
+rfstr.prop {{ _ \/ lp:Type }} F :- rfstr.prop Type F.
+rfstr.prop {{ lp:Type \/ _ }} F :- rfstr.prop Type F.
+rfstr.prop {{ is_true lp:Type }} F :- rfstr.bool Type F.
+rfstr.prop {{ @eq bool _ lp:Type }} F :- rfstr.bool Type F.
+rfstr.prop {{ @eq bool lp:Type _ }} F :- rfstr.bool Type F.
+rfstr.prop {{ @eq lp:Ty _ _ }} F :- field-mode, rf->sort F Ty, F = field _.
+rfstr.prop {{ @eq lp:Ty _ _ }} R :- not field-mode, rf->sort R Ty, R = ring _.
+
+pred rfstr.hyps i:list prop, o:rf.
+rfstr.hyps [decl _ _ H|_] F :- rfstr.prop H F.
+rfstr.hyps [_|Ctx] F :- rfstr.hyps Ctx F.
+
+pred rfstr i:list prop, i:term, o:rf.
+rfstr _ Type F :- field-mode => rfstr.prop Type F.
+rfstr _ Type R :- rfstr.prop Type R.
+rfstr Ctx _ F :- field-mode => rfstr.hyps Ctx F.
+rfstr Ctx _ R :- rfstr.hyps Ctx R.
+rfstr _ _ _ :- coq.ltac.fail _ "Cannot find a realDomainType".
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+% Reification procedure
+
+% [quote.cstr V In OutM Out] reifies integer constants
+% - [V] is a [ringType]
+% - [In] is a term of type [V],
+% - [OutM] is a reified constant of type [MExpr V]
+% - [Out] is a reified constant of type [PExpr Q]
+% - [OutQ] the reified [Out] as a [Q] (for <= 8.15 compat)
+pred quote.cstr i:term, i:term, o:term, o:term, o:term.
+quote.cstr V {{ GRing.zero lp:Z }} OutM Out OutQ :- ring->zmod V Z,
```

Comment:
I thought we should put a bang `!` after `ring->zmod` but not entirely sure. @gares

Reply:
Yes, it looks fine here (I guess it's equivalent as the remaining should never fail in practice)

Reply:
I agree that the behavior is the same, but putting a bang may avoid keeping some useless information about backtracking (but again, I'm not entirely sure).